tendermint/src/rpc/event_listener.rs

```diff
@@ -0,0 +1,166 @@
+//! Tendermint Websocket event listener client
+
+use crate::{
+    net,
+    rpc::Request,
+    rpc::{endpoint::subscribe, Error as RPCError},
+};
+use async_tungstenite::{tokio::connect_async, tokio::TokioAdapter, tungstenite::Message};
+use futures::prelude::*;
+use serde::{Deserialize, Serialize};
+
+use tokio::net::TcpStream;
+
+/// Event Listener over webocket. https://docs.tendermint.com/master/rpc/#/Websocket/subscribe
+pub struct EventListener {
+    socket: async_tungstenite::WebSocketStream<TokioAdapter<TcpStream>>,
+}
+
+impl EventListener {
+    /// Constructor for event listener
+    pub async fn connect(address: net::Address) -> Result<EventListener, RPCError> {
+        let (host, port) = match address {
+            net::Address::Tcp { host, port, .. } => (host, port),
+            other => {
+                return Err(RPCError::invalid_params(&format!(
+                    "invalid RPC address: {:?}",
+                    other
+                )));
+            }
+        };
+        //TODO This doesn't have any way to handle a connection over TLS
+        let (ws_stream, _unused_tls_stream) =
+            connect_async(&format!("ws://{}:{}/websocket", host, port)).await?;
+        Ok(EventListener { socket: ws_stream })
+    }
+    //TODO Have a query type instead of a string
+    /// Subscribe to event query stream over the websocket
+    pub async fn subscribe(&mut self, query: &str) -> Result<(), RPCError> {
+        self.socket
+            .send(Message::text(
+                subscribe::Request::new(query.to_owned()).into_json(),
+            ))
+            .await?;
+        Ok(())
+    }
+
+    /// Get the next event from the websocket
+    pub async fn get_event(&mut self) -> Result<Event, RPCError> {
+        let msg = self
+            .socket
+            .next()
+            .await
+            .ok_or_else(|| RPCError::websocket_error("web socket closed"))??;
+
+        match serde_json::from_str::<JSONRPC>(&msg.to_string()) {
+            Ok(data) => Ok(Event::JsonRPCTransactionResult {
+                data: Box::new(data),
+            }),
+
+            Err(_) => match msg.to_string().parse::<serde_json::Value>() {
+                Ok(data) => Ok(Event::GenericJSONEvent { data }),
+                Err(_) => Ok(Event::GenericStringEvent {
+                    data: msg.to_string(),
+                }),
+            },
+        }
+    }
+}
+
+//TODO more event types
+/// The Event enum is typed events emitted by the Websockets
+#[derive(Debug, Clone)]
+pub enum Event {
+    /// The result of the ABCI app processing a transaction, serialized as JSON RPC response
+    JsonRPCTransactionResult {
+        /// the tx result data
+        data: Box<JSONRPC>,
```

Comment:
Does this really need to contain the wrapped jsonrpc result?

Reply:
`data: JSONRPC,` was causing this clippy error: https://rust-lang.github.io/rust-clippy/master/index.html#large_enum_variant

Reply:
That's a neat trick, but I was wondering if data shouldn't rather be the events instead of the whole wrapped json result (independent of boxing or not).

Reply:
This represents an interesting question. Does the consumer want the raw data, the RPC response, just the Event? 

I guess my POV was provide everything and let the consumer figure it out.